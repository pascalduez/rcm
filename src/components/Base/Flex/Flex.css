.flex {
  display: flex;
}

.flexInline {
  display: inline-flex;
}


// -------------------- Direction: row -----------------------------------------

.flexRow {
  flex-direction: row;
}

.flexRowReverse {
  flex-direction: row-reverse;
}


// -------------------- Direction: column --------------------------------------

.flexCol {
  flex-direction: column;
}

.flexColReverse {
  flex-direction: column-reverse;
}


// -------------------- Wrap ---------------------------------------------------

.flexWrap {
  flex-wrap: wrap;
}

.flexNoWrap {
  flex-wrap: nowrap;
}

.flexWrapReverse {
  flex-wrap: wrap-reverse;
}


// -------------------- Justify: content ---------------------------------------

//
// Align items along the main axis of the current line of the flex container.
//

.flexJustifyStart {
  justify-content: flex-start;
}

.flexJustifyEnd {
  justify-content: flex-end;
}

.flexJustifyCenter {
  justify-content: center;
}

.flexJustifyBetween {
  justify-content: space-between;
}

.flexJustifyAround {
  justify-content: space-around;
}


// -------------------- Align: items -------------------------------------------

//
// Align items in the cross axis of the current line of the flex container
// Similar to `justify-content` but in the perpendicular direction
//

.flexAlignItemsStart {
  align-items: flex-start;
}

.flexAlignItemsEnd {
  align-items: flex-end;
}

.flexAlignItemsCenter {
  align-items: center;
}

.flexAlignItemsStretch {
  align-items: stretch;
}

.flexAlignItemsBaseline {
  align-items: baseline;
}


// -------------------- Align: content -----------------------------------------

//
// Aligns items within the flex container when there is extra
// space in the cross-axis
//
// Has no effect when there is only one line of flex items.
//

.flexAlignContentStart {
  align-content: flex-start;
}

.flexAlignContentEnd {
  align-content: flex-end;
}

.flexAlignContentCenter {
  align-content: center;
}

.flexAlignContentStretch {
  align-content: stretch;
}

.flexAlignContentBetween {
  align-content: space-between;
}

.flexAlignContentAround {
  align-content: space-around;
}


// -------------------- Applies to flex items ----------------------------------

//
// Override default alignment of single item when specified by `align-items`
//

.flexAlignSelfStart {
  align-self: flex-start;
}

.flexAlignSelfEnd {
  align-self: flex-end;
}

.flexAlignSelfCenter {
  align-self: center;
}

.flexAlignSelfStretch {
  align-self: stretch;
}

.flexAlignSelfBaseline {
  align-self: baseline;
}

.flexAlignSelfLastBaseline {
  align-self: last baseline;
}

.flexAlignSelfAuto {
  align-self: auto;
}


//
// Change order without editing underlying HTML
//

.flexOrderFirst {
  order: -1;
}

.flexOrderLast {
  order: 1;
}

.flexOrderNone {
  order: 0;
}


//
// Specify the flex grow factor, which determines how much the flex item will
// grow relative to the rest of the flex items in the flex container.
//
// Supports 1-5 proportions
//
// 1. Provide all values to avoid IE10 bug with shorthand flex
//    - http://git.io/vllC7
//
//    Use `0%` to avoid bug in IE10/11 with unitless flex basis. Using this
//    instead of `auto` as this matches what the default would be with `flex`
//    shorthand - http://git.io/vllWx
//

.flexGrow1 {
  flex: 1 1 0%; /* 1 */
}

.flexGrow2 {
  flex: 2 1 0%;
}

.flexGrow3 {
  flex: 3 1 0%;
}

.flexGrow4 {
  flex: 4 1 0%;
}

.flexGrow5 {
  flex: 5 1 0%;
}

//
// Specify the flex shrink factor, which determines how much the flex item will
// shrink relative to the rest of the flex items in the flex container.
//

.flexShrink0 {
  flex-shrink: 0;
}

.flexShrink1 {
  flex-shrink: 1;
}

.flexShrink2 {
  flex-shrink: 2;
}

.flexShrink3 {
  flex-shrink: 3;
}

.flexShrink4 {
  flex-shrink: 4;
}

.flexShrink5 {
  flex-shrink: 5;
}


//
// Aligning with `auto` margins
// http://www.w3.org/TR/css-flexbox-1/#auto-margins
//

.flexExpand {
  margin: auto;
}

.flexExpandLeft {
  margin-left: auto;
}

.flexExpandRight {
  margin-right: auto;
}

.flexExpandTop {
  margin-top: auto;
}

.flexExpandBottom {
  margin-bottom: auto;
}


//
// Basis
//

.flexBasisAuto {
  flex-basis: auto;
}

.flexBasis0 {
  flex-basis: 0;
}


//
// Shorthand
//
// Declares all values instead of keywords like 'initial' to work around IE10
// https://www.w3.org/TR/css-flexbox-1/#flex-common
//
// 1. Fixes issue in IE 10 where flex-basis is ignored - https://git.io/vllMt
//    This ensures it overrides flex-basis set in other utilities.
//

//
// Sizes the item based on the width/height properties
//

.flexInitial {
  flex: 0 1 auto;
  flex-basis: auto; // 1
}


//
// Sizes the item based on the width/height properties, but makes them fully
// flexible, so that they absorb any free space along the main axis.
//

.flexAuto {
  flex: 1 1 auto;
  flex-basis: auto; // 1
}


//
// Sizes the item according to the width/height properties, but makes the flex
// item fully inflexible. Similar to initial, except that flex items are
// not allowed to shrink, even in overflow situations.
//

.flexNone {
  flex: 0 0 auto;
  flex-basis: auto; // 1
}
