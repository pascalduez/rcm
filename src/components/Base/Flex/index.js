// @flow

export { default as withFlex } from './withFlex';
export type { Props as withFlexProps } from './withFlex';
export { default as Flex } from './Flex';
export type { Props as FlexProps } from './Flex';
