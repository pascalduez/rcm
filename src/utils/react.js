// @flow

import type { Element } from 'react';

export const getElementDisplayName = (element: Element<*>): string =>
  element.type.displayName ||
  // Function name.
  element.type.name ||
  // Function without a name.
  (typeof element.type === 'function' ? 'No Display Name' : element.type);
