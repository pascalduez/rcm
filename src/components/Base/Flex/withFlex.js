// @flow

import * as React from 'react';
import classNames from 'classnames';

import styles from './Flex.css';

export type Props = {
  flex?: boolean,
  flexInline?: boolean,
  flexRow?: boolean,
  flexRowReverse?: boolean,
  flexCol?: boolean,
  flexColReverse?: boolean,
  flexWrap?: boolean,
  flexNoWrap?: boolean,
  flexWrapReverse?: boolean,
  flexJustifyStart?: boolean,
  flexJustifyEnd?: boolean,
  flexJustifyCenter?: boolean,
  flexJustifyBetween?: boolean,
  flexJustifyAround?: boolean,
  flexAlignItemsStart?: boolean,
  flexAlignItemsEnd?: boolean,
  flexAlignItemsCenter?: boolean,
  flexAlignItemsStretch?: boolean,
  flexAlignItemsBaseline?: boolean,
  flexAlignContentStart?: boolean,
  flexAlignContentEnd?: boolean,
  flexAlignContentCenter?: boolean,
  flexAlignContentStretch?: boolean,
  flexAlignContentBetween?: boolean,
  flexAlignContentAround?: boolean,
  flexAlignSelfStart?: boolean,
  flexAlignSelfEnd?: boolean,
  flexAlignSelfCenter?: boolean,
  flexAlignSelfStretch?: boolean,
  flexAlignSelfBaseline?: boolean,
  flexAlignSelfLastBaseline?: boolean,
  flexAlignSelfAuto?: boolean,
  flexOrderFirst?: boolean,
  flexOrderLast?: boolean,
  flexOrderNone?: boolean,
  flexGrow1?: boolean,
  flexGrow2?: boolean,
  flexGrow3?: boolean,
  flexGrow4?: boolean,
  flexGrow5?: boolean,
  flexShrink0?: boolean,
  flexShrink1?: boolean,
  flexShrink2?: boolean,
  flexShrink3?: boolean,
  flexShrink4?: boolean,
  flexShrink5?: boolean,
  flexExpand?: boolean,
  flexExpandLeft?: boolean,
  flexExpandRight?: boolean,
  flexExpandTop?: boolean,
  flexExpandBottom?: boolean,
  flexBasisAuto?: boolean,
  flexBasis0?: boolean,
  flexInitial?: boolean,
  flexAuto?: boolean,
  flexNone?: boolean,
  className?: string,
};

/**
 * Helper: flexbox styling component.
 */
export default function withFlex(
  Comp: React.ComponentType<*>
): React.ComponentType<*> {
  function Flex(props: Props) {
    const {
      flex,
      flexInline,
      flexRow,
      flexRowReverse,
      flexCol,
      flexColReverse,
      flexWrap,
      flexNoWrap,
      flexWrapReverse,
      flexJustifyStart,
      flexJustifyEnd,
      flexJustifyCenter,
      flexJustifyBetween,
      flexJustifyAround,
      flexAlignItemsStart,
      flexAlignItemsEnd,
      flexAlignItemsCenter,
      flexAlignItemsStretch,
      flexAlignItemsBaseline,
      flexAlignContentStart,
      flexAlignContentEnd,
      flexAlignContentCenter,
      flexAlignContentStretch,
      flexAlignContentBetween,
      flexAlignContentAround,
      flexAlignSelfStart,
      flexAlignSelfEnd,
      flexAlignSelfCenter,
      flexAlignSelfStretch,
      flexAlignSelfBaseline,
      flexAlignSelfLastBaseline,
      flexAlignSelfAuto,
      flexOrderFirst,
      flexOrderLast,
      flexOrderNone,
      flexGrow1,
      flexGrow2,
      flexGrow3,
      flexGrow4,
      flexGrow5,
      flexShrink0,
      flexShrink1,
      flexShrink2,
      flexShrink3,
      flexShrink4,
      flexShrink5,
      flexExpand,
      flexExpandLeft,
      flexExpandRight,
      flexExpandTop,
      flexExpandBottom,
      flexBasisAuto,
      flexBasis0,
      flexInitial,
      flexAuto,
      flexNone,
      className,
      ...extraProps
    } = props;

    const rootClasses = classNames(
      {
        [styles.flex]: flex,
        [styles.flexInline]: flexInline,
        [styles.flexRow]: flexRow,
        [styles.flexRowReverse]: flexRowReverse,
        [styles.flexCol]: flexCol,
        [styles.flexColReverse]: flexColReverse,
        [styles.flexWrap]: flexWrap,
        [styles.flexNoWrap]: flexNoWrap,
        [styles.flexWrapReverse]: flexWrapReverse,
        [styles.flexJustifyStart]: flexJustifyStart,
        [styles.flexJustifyEnd]: flexJustifyEnd,
        [styles.flexJustifyCenter]: flexJustifyCenter,
        [styles.flexJustifyBetween]: flexJustifyBetween,
        [styles.flexJustifyAround]: flexJustifyAround,
        [styles.flexAlignItemsStart]: flexAlignItemsStart,
        [styles.flexAlignItemsEnd]: flexAlignItemsEnd,
        [styles.flexAlignItemsCenter]: flexAlignItemsCenter,
        [styles.flexAlignItemsStretch]: flexAlignItemsStretch,
        [styles.flexAlignItemsBaseline]: flexAlignItemsBaseline,
        [styles.flexAlignContentStart]: flexAlignContentStart,
        [styles.flexAlignContentEnd]: flexAlignContentEnd,
        [styles.flexAlignContentCenter]: flexAlignContentCenter,
        [styles.flexAlignContentStretch]: flexAlignContentStretch,
        [styles.flexAlignContentBetween]: flexAlignContentBetween,
        [styles.flexAlignContentAround]: flexAlignContentAround,
        [styles.flexAlignSelfStart]: flexAlignSelfStart,
        [styles.flexAlignSelfEnd]: flexAlignSelfEnd,
        [styles.flexAlignSelfCenter]: flexAlignSelfCenter,
        [styles.flexAlignSelfStretch]: flexAlignSelfStretch,
        [styles.flexAlignSelfBaseline]: flexAlignSelfBaseline,
        [styles.flexAlignSelfLastBaseline]: flexAlignSelfLastBaseline,
        [styles.flexAlignSelfAuto]: flexAlignSelfAuto,
        [styles.flexOrderFirst]: flexOrderFirst,
        [styles.flexOrderLast]: flexOrderLast,
        [styles.flexOrderNone]: flexOrderNone,
        [styles.flexGrow1]: flexGrow1,
        [styles.flexGrow2]: flexGrow2,
        [styles.flexGrow3]: flexGrow3,
        [styles.flexGrow4]: flexGrow4,
        [styles.flexGrow5]: flexGrow5,
        [styles.flexShrink0]: flexShrink0,
        [styles.flexShrink1]: flexShrink1,
        [styles.flexShrink2]: flexShrink2,
        [styles.flexShrink3]: flexShrink3,
        [styles.flexShrink4]: flexShrink4,
        [styles.flexShrink5]: flexShrink5,
        [styles.flexExpand]: flexExpand,
        [styles.flexExpandLeft]: flexExpandLeft,
        [styles.flexExpandRight]: flexExpandRight,
        [styles.flexExpandTop]: flexExpandTop,
        [styles.flexExpandBottom]: flexExpandBottom,
        [styles.flexBasisAuto]: flexBasisAuto,
        [styles.flexBasis0]: flexBasis0,
        [styles.flexInitial]: flexInitial,
        [styles.flexAuto]: flexAuto,
        [styles.flexNone]: flexNone,
      },
      className
    );

    return <Comp className={rootClasses} {...extraProps} />;
  }

  return Flex;
}
