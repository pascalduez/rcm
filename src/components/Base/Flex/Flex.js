// @flow

import * as React from 'react';

import withFlex, { type Props as InjectedProps } from './withFlex';

export type Props = {
  tagName: string,
  className?: string,
  children?: React.Node,
} & InjectedProps;

function Flex({ tagName = 'div', children, ...extraProps }: Props) {
  return React.createElement(tagName, extraProps, children);
}

const Typed = withFlex(Flex);

Flex.displayName = 'Flex';
Typed.displayName = 'Flex';

export default Typed;
