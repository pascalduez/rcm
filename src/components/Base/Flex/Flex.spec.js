// @flow

import * as React from 'react';
import { render } from 'enzyme';

import Flex from './Flex';

describe('<Flex />', () => {
  it('should render the proper default markup and classes', () => {
    const wrapper = render(<Flex />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render the proper markup following the `tagName` property', () => {
    const wrapper = render(<Flex tagName="span" />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should add the proper classes based on props', () => {
    const wrapper = render(
      <Flex flex flexJustifyCenter flexAlignItemsCenter />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should pass custom classes trough', () => {
    const wrapper = render(<Flex className="test" />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should pass extra props through', () => {
    const wrapper = render(
      <Flex style={{ display: 'block' }}>Just a text</Flex>
    );

    expect(wrapper).toMatchSnapshot();
  });
});
