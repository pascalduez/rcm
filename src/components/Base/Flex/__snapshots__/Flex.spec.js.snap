// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Flex /> should add the proper classes based on props 1`] = `
<div
  class=""
/>
`;

exports[`<Flex /> should pass custom classes trough 1`] = `
<div
  class="test"
/>
`;

exports[`<Flex /> should pass extra props through 1`] = `
<div
  class=""
  style="display:block"
>
  Just a text
</div>
`;

exports[`<Flex /> should render the proper default markup and classes 1`] = `
<div
  class=""
/>
`;

exports[`<Flex /> should render the proper markup following the \`tagName\` property 1`] = `
<span
  class=""
/>
`;
